src/components/animation/ripple.ts

```diff
@@ -0,0 +1,220 @@
+
+const RIPPLE_KEY_FRAMES = [
+  { transform: "scale(0)" },
+  { transform: "scale(0.8)", offset: 0.33 },
+  { opacity: "0", transform: "scale(1)", offset: 1 },
+];
+
+css/* scss */ `
+  .o-ripple {
+    z-index: 1;
+  }
+`;
+
+interface RippleProps {
+  color: string;
+  opacity: number;
+  duration: number;
+
+  /** If true, the ripple will play from the element center instead of the position of the click */
+  ignoreClickPosition?: boolean;
+
+  /** Width of the ripple. Defaults to the width of the element the ripple is on (without margins). */
+  width?: number;
+  /** Height of the ripple. Defaults to the height of the element the ripple is on (without margins). */
+  height?: number;
+
+  offsetY?: number;
+  offsetX?: number;
+
+  allowOverflow?: boolean;
+  enabled: boolean;
+  onAnimationEnd: () => void;
+  class: string;
+}
+
+interface RippleEffectProps
+  extends Omit<Required<RippleProps>, "ignoreClickPosition" | "enabled" | "class"> {
+  x: string;
+  y: string;
+}
+
+interface RippleDef {
+  rippleRect: Rect | undefined;
+  id: number;
+}
+
+interface RippleState {
+  ripples: Array<RippleDef>;
+}
+
+interface RectWithMargins extends Rect {
+  marginTop: number;
+  marginLeft: number;
+}
+
+class RippleEffect extends Component<RippleEffectProps, SpreadsheetChildEnv> {
+  static template = "o-spreadsheet-RippleEffect";
+  private rippleRef = useRef("ripple");
+
+  setup() {
+    onMounted(() => {
+      const rippleEl = this.rippleRef.el;
+      if (!rippleEl || !rippleEl.animate) return;
+      const animation = rippleEl.animate(RIPPLE_KEY_FRAMES, {
+        duration: this.props.duration,
+        easing: "ease-out",
+      });
+      animation.onfinish = () => this.props.onAnimationEnd();
+    });
+  }
+
+  get rippleStyle() {
+    const { x, y, width, height } = this.props;
+    const offsetX = this.props.offsetX || 0;
+    const offsetY = this.props.offsetY || 0;
+    return cssPropertiesToCss({
+      transform: "scale(0)",
+      left: x,
+      top: y,
+      "margin-left": `${-width / 2 + offsetX}px`,
+      "margin-top": `${-height / 2 + offsetY}px`,
+      width: `${width}px`,
+      height: `${height}px`,
+      background: this.props.color,
+      "border-radius": "100%",
+      opacity: `${this.props.opacity}`,
+    });
+  }
+}
+
+RippleEffect.props = {
+  x: String,
+  y: String,
+  color: String,
+  opacity: Number,
+  duration: Number,
+  width: Number,
+  height: Number,
+  offsetY: Number,
+  offsetX: Number,
+  allowOverflow: Boolean,
+  onAnimationEnd: Function,
+};
+
+export class Ripple extends Component<RippleProps, SpreadsheetChildEnv> {
+  static template = "o-spreadsheet-Ripple";
+  static components = { RippleEffect };
+  static defaultProps = {
+    color: "#aaaaaa",
+    opacity: 0.4,
+    duration: 800,
+    enabled: true,
+    onAnimationEnd: () => {},
+    class: "",
+  };
+
+  private childContainer = useRef("childContainer");
+
+  private state = useState<RippleState>({ ripples: [] });
```

Comment:
why is there mutliple ripples?

Reply:
'cause you can click multiple times, and we will have 1 ripple per click